amgcl/preconditioner/cpr.hpp

```diff
@@ -153,264 +151,270 @@ class cpr {
             S = std::make_shared<SPrecond>(K, prm.sprecond, bprm);
         }
 
+        template <class Matrix>
+        void update_pressure_system(
+                const Matrix &K,
+                const backend_params &bprm = backend_params())
+        {
+          init_pressure_system(std::make_shared<build_matrix>(K), bprm,
+            std::integral_constant<bool, math::static_rows<value_type>::value == 1>());
+        }
+
     private:
         size_t n, np;
 
         std::shared_ptr<PPrecond> P;
         std::shared_ptr<SPrecond> S;
 
-        std::shared_ptr<matrix_p> Fpp, Scatter;
+        std::shared_ptr<matrix_p> App, Fpp, Scatter;
```

Comment:
The `App` matrix is stored here as `matrix_p`, which is, in general, the backend matrix with opaque storage that only supports spmv operation, and is created after the setup is complete. But in the new [`init_pressure_system`](https://github.com/ddemidov/amgcl/pull/117/files#diff-64b00f6354836a3af547df98126c4132R200) it is allocated as `build_matrix` (should it be `build_matrix_p`?) which is used during setup. For the builtin backend these are the same types, but this will fail for any other backend.

Another note on storing `App` as the class member: the pressure preconditioner P does that anyway (App is converted to `matrix_p` after the setup of `PPrecond` is complete and may be obtained via P->system_matrix()). So may be it would be better to make `init_pressure_system` to return `App` as `shared_ptr<build_matrix_p>`?

Reply:
Thank you for the comments - that makes a lot of sense. I will fix the type issues and let `init_pressure_system` return the pressure matrix instead.